### PR TITLE
Implement table widget for Ecto queries

### PR DESCRIPTION
**Pull request by jonatanklosko**
Mostly useful with Kino as a dev dependency in Mix projects, but may also be useful for Ecto tutorial notebooks.

https://user-images.githubusercontent.com/17034772/126636714-45b7495a-31a0-4400-9242-d7f60b9ed81a.mp4

### CODE REVIEW

**Comment by jonatanklosko** [comment]
[Here](https://gist.github.com/jonatanklosko/5dfedc5bfd88c3143ce01a19192b574e) is a self-contained example notebook.

File: lib/kino/ecto.ex [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+        query
+      end
+
+    records = repo.all(query)
+
+    {count, records}
+  end
+
+  defp default_select_query?(queryable) do
+    query = from(q in queryable, [])
```
Comment (josevalim) [inline lib/kino/ecto.ex:149]: Call to_queryable instead?

File: lib/kino/utils/table.ex [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+    # Note: id is opaque to the client, and we don't need it for now
+    %{id: nil, fields: fields}
+  end
+
+  @doc """
+  Checks if the given term is an Ecto.Schema.
+  """
+  def ecto_schema?(queryable) do
+    is_atom(queryable) and function_exported?(queryable, :__schema__, 1)
```
Comment (josevalim) [inline lib/kino/utils/table.ex:119]: Maybe it is better to convert to a queryable and then look under the from?

File: lib/kino/utils/table.ex [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+      end)
+
+    # Note: id is opaque to the client, and we don't need it for now
+    %{id: nil, fields: fields}
+  end
+
+  @doc """
+  Checks if the given term is an Ecto.Schema.
+  """
+  def ecto_schema?(queryable) do
```
Comment (josevalim) [inline lib/kino/utils/table.ex:118]: Is this reused anywhere? If not, maybe we should move to the Ecto module?

File: lib/kino/utils/table.ex [hunk 4]
```diff
@@ -0,0 +1,121 @@
+defmodule Kino.Utils.Table do
```
Comment (josevalim) [inline lib/kino/utils/table.ex:1]: Maybe Kino.Table.Utils? It feels Table is more important but that’s a nitpick. :)

File: lib/kino/utils/table.ex (hunk 2 above)
Comment (jonatanklosko) [reply lib/kino/utils/table.ex:119]: I may be missing something, but the information doesn't seem to be explicitly there:

image

File: lib/kino/utils/table.ex (hunk 3 above)
Comment (jonatanklosko) [reply lib/kino/utils/table.ex:118]: It's used in this module in `columns_for_record`.

File: lib/kino/ecto.ex [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+  end
+
+  def handle_info({:DOWN, ref, :process, _object, _reason}, %{parent_monitor_ref: ref} = state) do
+    {:stop, :shutdown, state}
+  end
+
+  defp get_records(repo, queryable, rows_spec) do
+    count = repo.aggregate(queryable, :count)
+
+    query = from(q in queryable, limit: ^rows_spec.limit, offset: ^rows_spec.offset)
```
Comment (josevalim) [inline lib/kino/ecto.ex:133]: Should we check the query has no limit and offset upfront? Otherwise this will raise. Alternatively we can call Query.exclude and remove both values upfront.

File: lib/kino/ecto.ex [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+        {"Query", []}
+      end
+
+    features =
+      Kino.Utils.truthy_keys(
+        refetch: true,
+        pagination: true,
+        # If the user specifies custom select, the record keys
+        # are not valid "order by" fields, so we disable sorting
+        sorting: default_select_query?(state.queryable)
```
Comment (josevalim) [inline lib/kino/ecto.ex:95]: Should we sort if there is already sorting?

File: lib/kino/ecto.ex (hunk 5 above)
Comment (jonatanklosko) [reply lib/kino/ecto.ex:133]: Does it raise thought? It looks like the values are just overridden:

image

File: lib/kino/ecto.ex (hunk 6 above)
Comment (jonatanklosko) [reply lib/kino/ecto.ex:95]: I should be fine. We show the initially sorted results, but the user can also click on columns to override the sorting.

File: lib/kino/utils/table.ex (hunk 4 above)
Comment (jonatanklosko) [reply lib/kino/utils/table.ex:1]: My reasoning is that `Utils` is less specific and can thus serve as a namespace, no strong opinion though.

Comment (josevalim) [reply lib/kino/utils/table.ex:1]: Fine by me :)

File: lib/kino/ecto.ex (hunk 6 above)
Comment (josevalim) [reply lib/kino/ecto.ex:95]: The issue is that we are not overriding, we are adding to the sorting. We would need to exclude the existing sorting first. We should have tests for this :)

Comment (jonatanklosko) [reply lib/kino/ecto.ex:95]: Oh right! Exclude sounds good. Is there some test Repo that we could use or how would you go about the tests?

**Review by josevalim — COMMENTED** [review]
Comments added. We should probably devise a testing scheme where we emulate the messages to handle some of the corner cases.

File: lib/kino/ecto.ex (hunk 6 above)
Comment (josevalim) [reply lib/kino/ecto.ex:95]: There is an ETSO adapter for ETS, no dependencies but I am not sure if it supports order by. Otherwise either PG or mocks.

File: lib/kino/ecto.ex (hunk 5 above)
Comment (josevalim) [reply lib/kino/ecto.ex:133]: Ah, I thought we raised. Good then!

File: lib/kino/utils/table.ex (hunk 3 above)
Comment (josevalim) [reply lib/kino/utils/table.ex:118]: Hrm, not sure if it is worth having it there. It is like the handling of Ecto queries is now everywhere instead of isolated in a single place. Consider splitting that apart or having Ecto not call it?

Also remember that Ecto’s select can return a tuple, so we need to take that into account too.

Comment (jonatanklosko) [reply lib/kino/utils/table.ex:118]: This was kinda intentional, because if someone passes a list of Ecto structs into `DataTable`, we may as well use the fields information it provides, right?

> Also remember that Ecto’s select can return a tuple, so we need to take that into account too.

Yeah, it can be a map or a single value too, that's the reason why I moved much stuff to `Table.Utils`, because it's pretty much the same for `DataTable` and `ETS`.

Comment (josevalim) [reply lib/kino/utils/table.ex:118]: > This was kinda intentional, because if someone passes a list of Ecto structs into DataTable, we may as well use the fields information it provides, right?

i would just treat them as structs since the fields are in the struct too? What benefits would be added by looking into Ecto?

Comment (jonatanklosko) [reply lib/kino/utils/table.ex:118]: The benefits are that the columns are ordered as per the schema definition, which should feel more natural. Also this way we don't show virtual fields such as associations.

Comment (josevalim) [reply lib/kino/utils/table.ex:118]: Hrm, I see. Sounds good to me then!

File: lib/kino/utils/table.ex (hunk 2 above)
Comment (jonatanklosko) [reply lib/kino/utils/table.ex]: Ah, as far as I can see the second element in source is either a schema or `nil`, so this helps. I updated the extraction in ca7064448bf3f42f05374a15484f6c4cee52852c.

File: lib/kino/utils/table.ex [hunk 7]
```diff
@@ -113,9 +113,27 @@ defmodule Kino.Utils.Table do
...
-  def ecto_schema?(queryable) do
-    is_atom(queryable) and function_exported?(queryable, :__schema__, 1)
+  def ecto_schema(queryable)
+
+  def ecto_schema(struct) when is_struct(struct) do
+    ecto_schema(struct.__struct__)
+  end
+
+  def ecto_schema(queryable) when is_atom(queryable) do
+    if function_exported?(queryable, :__schema__, 1) do
```
Comment (josevalim) [inline lib/kino/utils/table.ex:127]:
```suggestion
    if Code.ensure_loaded?(queryable) and function_exported?(queryable, :__schema__, 1) do
```

File: lib/kino/utils/table.ex [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+      queryable
+    else
+      nil
+    end
+  end
```
Comment (jonatanklosko) [inline lib/kino/utils/table.ex:132]: We could use `Ecto.Queryable.to_query`, but we don't want this function to rely on Ecto being installed.

File: lib/kino/ecto.ex (hunk 6 above)
Comment (jonatanklosko) [reply lib/kino/ecto.ex:89]: There's an old [pending PR](https://github.com/evadne/etso/pull/12) that adds ORDER BY support ot ETSO. Either way, to make precise assertions we need mocking in some form. I've just introduced a `MockRepo` that we can assert and resolve calls against, let me know if it makes sense :)

**Review by josevalim — APPROVED** [review]
